helix-core/src/main/java/org/apache/helix/manager/zk/ZkBaseDataAccessor.java

```diff
@@ -92,19 +98,52 @@ public ZkBaseDataAccessor(HelixZkClient zkClient) {
     _zkClient = zkClient;
   }
 
+  /**
+   * Lazy initialization is used to reduce the cost of unnecessary opened zkConnections
+   * @return singleton instance of non-ZnRecord ZkClient
```

Comment:
Here's another idea. If keeping many active ZkConnections is expensive (and how? memory intensive?), could we close the connection after each operation is done, and connect the ZkClient right before the start of an operation?

Reply:
I think the option has been discussed last week and I have reasons against it: worse case it will open/close many zk connections and each round of operation is I/O expensive. Performance is a problem and the cost of open/close zk connections. Compared with that, the one single additional zkConnection by lazy initialization at least only add one additional zkConnection and the performance is better.